dipy/tracking/vox2track.pyx

```diff
@@ -162,6 +161,172 @@ def streamline_mapping(streamlines, voxel_size=None, affine=None,
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
+cdef inline double norm(double x, double y, double z) nogil:
+    cdef double val = sqrt(x*x + y*y + z*z)
+    return val
+
+
+# Changing this to a memview was slower.
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline void c_get_closest_edge(double p_x, double p_y, double p_z,
+                                    double d_x, double d_y, double d_z,
+                                    cnp.double_t[:] edge,
+                                    double eps=1.) nogil:
+     edge[0] = floor(p_x + eps) if d_x >= 0.0 else ceil(p_x - eps)
+     edge[1] = floor(p_y + eps) if d_y >= 0.0 else ceil(p_y - eps)
+     edge[2] = floor(p_z + eps) if d_z >= 0.0 else ceil(p_z - eps)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def streamlines_in_mask(list streamlines,
+                        cnp.uint8_t[:,:,:] mask,
+                        lin, offset):
+    """Filters streamlines based on whether or not they pass through an ROI,
+    using a line-based algorithm for compressed tract.
+
+    Parameters
+    ----------
+    streamlines : sequence
+        A sequence of streamlines.
+    mask : array-like (uint8)
+        A mask used as a target. Non-zero values are considered to be within
+        the target region.
+    lin : array (3, 3)
+        Transpose of the linear part of the mapping to voxel space, (ie
+        ``inv(affine)[:3, :3].T``)
+    offset : array or scaler
+        Offset part of the mapping (ie, ``inv(affine)[:3, 3]``) + ``.5``. The
+        half voxel shift is so that truncating the result of this mapping
+        will give the correct integer voxel coordinate.
+
+    Returns
+    -------
+    in_mask : array of bool
+        0 if passing in mask, 1 otherwise
+    """
+    cdef cnp.double_t[:,:] voxel_indices
+
+    # Declaring it here and passing to friend function because creating them
+    # there would cost too much.
+    cdef cnp.double_t[:] current_pt = np.zeros(3, dtype=np.double)
+    cdef cnp.double_t[:] next_pt = np.zeros(3, dtype=np.double)
+    cdef cnp.double_t[:] direction = np.zeros(3, dtype=np.double)
+    cdef cnp.double_t[:] current_edge = np.zeros(3, dtype=np.double)
+
+    cdef int nb_streamlines = len(streamlines)
+    cdef cnp.uint8_t[:] in_mask = np.zeros(nb_streamlines, dtype=np.uint8)
+    cdef int track_idx
+
+    for track_idx in range(nb_streamlines):
+        # Can't call _to_voxel_coordinates because it casts to int
+        voxel_indices = np.dot(streamlines[track_idx], lin) + offset
+
+        if streamline_in_mask(voxel_indices, mask,
```

Comment:
Yes, I tried. It did it this way in my first version. I thought it would be clearer in 2 functions but wondered about the cost of the function calls. I do not remember if there was any significant difference but I'm gonna guess there wasn't or I wouldn't have created another function.